aten/src/ATen/native/NaiveDilatedConvolution.h

```diff
@@ -0,0 +1,432 @@
+#include <ATen/native/im2col.h>
+#include <ATen/native/vol2col.h>
+#include <ATen/Utils.h>
+#include <c10/util/accumulate.h>
+#include <c10/util/irange.h>
+#include <tuple>
+
+namespace at {
+namespace native {
+namespace {
+
+// hyper-volume to column, CPU
+template <typename Dtype, int64_t dim>
+void hvol2col(
+    const Dtype* data_hvol,
+    const int channels,
+    const IntArrayRef input_size,
+    const IntArrayRef output_size,
+    const IntArrayRef kernel_size,
+    const IntArrayRef stride_size,
+    const IntArrayRef pad_size,
+    const IntArrayRef dilation_size,
+    Dtype* data_col) {
+  if (dim == 3) {
+    vol2col<Dtype>(
+        data_hvol,
+        channels,
+        input_size[0],
+        input_size[1],
+        input_size[2],
+        output_size[0],
+        output_size[1],
+        output_size[2],
+        kernel_size[0],
+        kernel_size[1],
+        kernel_size[2],
+        pad_size[0],
+        pad_size[1],
+        pad_size[2],
+        stride_size[0],
+        stride_size[1],
+        stride_size[2],
+        dilation_size[0],
+        dilation_size[1],
+        dilation_size[2],
+        data_col);
+  }
+  if (dim == 2) {
+    im2col<Dtype>(
+        data_hvol,
+        channels,
+        input_size[0],
+        input_size[1],
+        output_size[0],
+        output_size[1],
+        kernel_size[0],
+        kernel_size[1],
+        pad_size[0],
+        pad_size[1],
+        stride_size[0],
+        stride_size[1],
+        dilation_size[0],
+        dilation_size[1],
+        data_col);
+  }
+}
+
+// column to hyper-volume, CPU
+template <typename Dtype, int64_t dim>
+void col2hvol(
+    const Dtype* data_col,
+    const int channels,
+    const IntArrayRef input_size,
+    const IntArrayRef output_size,
+    const IntArrayRef kernel_size,
+    const IntArrayRef stride_size,
+    const IntArrayRef pad_size,
+    const IntArrayRef dilation_size,
+    Dtype* data_hvol) {
+  if (dim == 3) {
+    col2vol<Dtype>(
+        data_col,
+        channels,
+        input_size[0],
+        input_size[1],
+        input_size[2],
+        output_size[0],
+        output_size[1],
+        output_size[2],
+        kernel_size[0],
+        kernel_size[1],
+        kernel_size[2],
+        pad_size[0],
+        pad_size[1],
+        pad_size[2],
+        stride_size[0],
+        stride_size[1],
+        stride_size[2],
+        dilation_size[0],
+        dilation_size[1],
+        dilation_size[2],
+        data_hvol);
+  }
+  if (dim == 2) {
+    col2im<Dtype>(
+        data_col,
+        channels,
+        input_size[0],
+        input_size[1],
+        output_size[0],
+        output_size[1],
+        kernel_size[0],
+        kernel_size[1],
+        pad_size[0],
+        pad_size[1],
+        stride_size[0],
+        stride_size[1],
+        dilation_size[0],
+        dilation_size[1],
+        data_hvol);
+  }
+}
+
+/*
+   check tensor data locations
+*/
+void slow_conv_dilated_location_check(
```

Comment:
Should be inline, otherwise there will be duplicate symbol linker errors if NaiveDilatedConvolution.h is included in more files

Reply:
Note that this is defined in an anonymous namespace, which I think avoids the multiply-defined problem.

Reply:
You're right, thanks for the catch!